Add aditional features to install_cmdstan

rok-cesnovar opened the pull request:
Closes #38 

This PR changes the behavior of `install_cmdstan` so that it will now only install `cmdstan` if there is a newer version available.
This PR also the following arguments to the `install_cmdstan`:
- backup_existing = FALSE
If TRUE will backup the existing installation if found in the same location. It will backup it to `cmdstan_backup`. If there is already a `cmdstan_backup` folder it gets removed.  (Should we stop in this case and let users delete it manually?).

-  force_reinstall = FALSE
Will reinstall cmdstan even if the version is already the latest

- github_repo_clone = FALSE
Clones the github repository Cmdstan and checks out the develop branch.

There is some other cleanup in `make_cmdstan.sh` I will comment directly in the code.

rok-cesnovar commented on inst/make_cmdstan.sh:
```diff
@@ -28,64 +53,115 @@ if [ -z ${RELDIR} ]; then
     RELDIR="$HOME/.cmdstanr"
 fi
 
-# if [[ -e ${RELDIR} && ! -d ${RELDIR} ]]; then
```
This was commented out so this can be safely removed.

rok-cesnovar commented on inst/make_cmdstan.sh:
```diff
@@ -28,64 +53,115 @@ if [ -z ${RELDIR} ]; then
-fi
+    pushd ${RELDIR} > /dev/null
+    #testing if there are files in the cmdstan folder
+    if [[ -d ${INSTALL_DIR} && -f ${INSTALL_DIR}/makefile ]]; then
```
I changed this to check for `cmdstan` and `cmdstan/makefile`. Looking for `bin/stanc` is not OK as on Windows it will be `bin/stanc.exe`.

rok-cesnovar commented on inst/make_cmdstan.sh:
```diff
@@ -28,64 +53,115 @@ if [ -z ${RELDIR} ]; then
+            echo "installed cmdstan is the latest version"
+            echo "only running clean and rebuild"
+            pushd cmdstan > /dev/null
+            echo "rebuilding cmdstan binaries"
```
If the installed cmdstan is already the latest version I made it clean-all and rebuild. This could be used in case of any weird issue as a first step in troubleshooting.

bbbales2 commented:
The force reinstall and github_repo_clone thing seem useful. I'm not sold on the automatic backup. Why not just print a message "cmdstan found at X. Please backup or remove this installation"? Force reinstall can be used for people who want to overwrite what's there, and this way we avoid deleting anything accidentally.

Re the Github thing, it might be handy to be able to specify git repo/branch compos. Like I've got a branch of cmdstan with stuff for Aki. If I could just say ```install_cmdstan(repo = "mygitrepo", branch = "foraki")``` that could make his life a little easier.

jgabry commented:
> @jgabry your thoughts?

Sounds good to me! Just one clarification: if we change the signature to allow specifying the repo and branch in `install_cmdstan()` do we also need a `cmdstan_checkout_git_branch()` function? Do they serve different purposes? Maybe I misunderstood and we do need both, which is totally fine, just curious.

bbbales2 commented:
> Its not something the general user base wont use all that much but its probably useful for cmdstan, algorithms & math developers.

I think this would get used a lot.

> would by default stop if it finds a cmdstan folder in the install location. And then the user would run it with install_cmdstan(overwrite=TRUE)

I like this.

> github_repo = "stan-dev/cmdstan"

Does it need to be github specific? Aren't these URLs generic: ```https://github.com/stan-dev/cmdstan.git```?

> cmdstan_checkout_git_branch(name, clean_and_rebuild = FALSE)

This seems kinda risky. If the use case is automated testing, I'd just assume not cleaning and rebuilding would be too misleading. I'll leave this to @jgabry. I think I'd lean no on this, but my opinion could change in an instant lol. I definitely expect to go into my cmdstanr cmdstan directories and manually do git stuff (I just don't see a reason to do that stuff from R). What if there were local changes or something? I guess I'm digging.

jgabry commented:
Thanks to you both! Good ideas and suggestions. 

> This seems kinda risky.

Yeah fair enough. Certainly the default should be to rebuild. 

> The first one clones and checks out a branch the other one just switches a branch.

Ah ok, I see, that make sense.

rok-cesnovar commented:
I pushed the following changes: 
- force_reinstall is now overwrite
- removed backing up,  install stops it if finds cmdstan/ and overwrite=FALSE
- added the cmdstan_git_checkout_branch ( I am open to changing the name here, dont really feel its a good name, but dont have a better one)

rok-cesnovar commented:
This should now also fix #53 though I am not seeing charts get posted here (codecov runs fine https://travis-ci.org/stan-dev/cmdstanr/jobs/605441375#L736).

jgabry commented:
Awesome, will take a look shortly!

jgabry commented:
@rok-cesnovar I just pushed a few changes, but in case there's a problem or I misunderstood the intent of something we can revert my changes if necessary. The main things I changed are:

* In `make_cmdstan.sh`: 
  - Define error_msg() so we can get red text for important messages
  - Error if overwrite=FALSE and there's an existing installation

* In `install_cmdstan()`: 
  - made it so it still returns the install_log even if there's an error
  - makes a different suggestion to user about an error depending on if `quiet` is TRUE or FALSE

* In `cmdstan_git_checkout_branch`
  - changed argument name `branch_name` to `repo_name` to match `install_cmdstan` (is there a reason to name them differently? maybe there is in which case we can change it back!) 
  
What do you think? Hopefully I didn't break anything making these changes.

bbbales2 commented on man/install_cmdstan.Rd:
```diff
@@ -4,26 +4,49 @@
+\item{overwrite}{Defaults to \code{FALSE}, but if \code{TRUE} will download and
+reinstall CmdStan even if an existing installation is found.}
+
+\item{repo_clone}{Defaults to \code{FALSE}, but if \code{TRUE} will install a
```
If \code{FALSE} installs latest CmdStan release from tarball. If \code{TRUE} installs a git clone of CmdStan and checks out the specified branch.

bbbales2 commented:
Thanks Rok! This looks cool. I pulled it down and tried to break it in various ways. There are a couple weird errors, but not sure they're real issues. The clean_and_rebuild flag doesn't seem to work though.

I got this error during a checkout:
```
g++ -pipe -pthread -D_REENTRANT -O3 -I stan/lib/stan_math/lib/sundials_4.1.0/include -DNO_FPRINTF_OUTPUT   -c -x c -include stan/lib/stan_math/lib/sundials_4.1.0/include/stan_sundials_printf_override.hpp stan/lib/stan_math/lib/sundials_4.1.0/src/kinsol/kinsol_spils.c -o stan/lib/stan_math/lib/sundials_4.1.0/src/kinsol/kinsol_spils.o
/bin/sh: 1: [[: not found
touch stan/lib/stan_math/lib/tbb/tbb-make-check
g++ -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare     -I stan/lib/stan_math/lib/tbb_2019_U8/include -O3 -I src -I stan/src -I lib/rapidjson_1.1.0/ -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.69.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include    -DBOOST_DISABLE_ASSERTS      -c -o src/cmdstan/main.o src/cmdstan/main.cpp
make: *** No rule to make target 'examples/bernoulli/bernoulli'.  Stop.
make: *** Waiting for unfinished jobs....
```

The command was:
```
install_cmdstan("/home/bbales2/tmp/tmp", cores = 8,
                repo_clone = TRUE,
                repo_url = "https://github.com/bbbales2/cmdstan-rus.git",
                repo_branch = "experimental/warmup")
```

The cmdstan seemed to work though.

I then went to check out another branch:
```
cmdstan_git_checkout_branch("experimental/warmup-auto", cores = 8, clean_and_rebuild = FALSE)
```

And cmdstan started rebuilding (even though clean_and_rebuild is false).

I got this error when trying to checkout when I had my cmdstan_path set to a Github cmdstan folder -- it seems formatted weirdly:

```
> cmdstan_git_checkout_branch("develop")
Running bash /home/bbales2/R/x86_64-pc-linux-gnu-library/3.6/cmdstanr/make_cmdstan.sh '-d /home/bbales2/.cmdstanr/cmdstan' -j2 -c -b develop
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
Error: pulling from the repository failed.
Please check that the branch exists in the repository to be cloned,
and that there are not any unstashed changes in  /home/bbales2/.cmdstanr/cmdstan/cmdstan.
$status
[1] 0

$stdout
[1] ""

$stderr
[1] "fatal: not a git repository (or any parent up to mount point /)\nStopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).\nError: pulling from the repository failed.\nPlease check that the branch exists in the repository to be cloned,\nand that there are not any unstashed changes in  /home/bbales2/.cmdstanr/cmdstan/cmdstan.\n"

$timeout
[1] FALSE
```

jgabry commented:
Thanks @bbbales2!

> it seems formatted weirdly 

I think at least the weird formatting was my fault (if you mean the list that it outputs after the messages). I’ll clean that up.

jgabry commented:
I think my latest commits are also causing a Travis error but I think I know how to fix that.

rok-cesnovar commented:
@bbbales2 Thanks for the thorough check!

> /bin/sh: 1: [[: not found

This one is a cmdstan issue I think we cleaned up on the main repo (or maybe not, not 100% sure). Its a problem that make calls some shell script and bash is fine with it but if you use sh it spits out this warning. However this doesnt stop building its just an ugly warning.
The issue here was that your repo doesnt have the bernoulli example and the default behavior was to build cmdstan and the bernoulli example. I removed the latter if installing a repo. Seems reasonable. And thank you again for testing to discover this one.

> The clean_and_rebuild flag doesn't seem to work though.

Yeah, handling of that flag was missing. Slipped through, sorry. Should be good now.

rok-cesnovar commented:
And thank you @jgabry for cleaning up my docs and prints. One of these day I will learn to write proper doc :)

jgabry commented:
Thanks @rok-cesnovar! @bbbales2 any follow ups?

bbbales2 commented:
Lemme check it real quick.

bbbales2 commented:
All looks good to me

jgabry commented:
Great, thanks!

jgabry commented:
> And thank you @jgabry for cleaning up my docs and prints. One of these day I will learn to write proper doc :)

@rok-cesnovar the doc wasn't bad! I just had some ideas to make it better (including changing somethings that I had previously documented, so not just yours!)
